kazuki-core-test/src/test/kotlin/com/anaplan/engineering/kazuki/core/test/TestSet.kt

```diff
@@ -1,87 +1,135 @@
 package com.anaplan.engineering.kazuki.core.test
 
 import com.anaplan.engineering.kazuki.core.*
+import com.anaplan.engineering.kazuki.core.Set1Extension_Module.mk_Set1Extension
+import com.anaplan.engineering.kazuki.core.SetExtension_Module.mk_SetExtension
+import org.junit.runner.RunWith
+import org.junit.runners.Parameterized
 import kotlin.test.Test
 import kotlin.test.assertEquals
-import kotlin.test.assertFailsWith
 
-class TestSet {
+
+@RunWith(Parameterized::class)
+class TestSet(
+    private val allowsEmpty: Boolean,
+    private val creator: (Collection<*>) -> Set<*>
+) {
+
+    companion object {
+        @JvmStatic
+        @Parameterized.Parameters
+        fun creators() =
+            listOf(
+                arrayOf(true, { m: Collection<*> -> mk_Set(*m.toTypedArray()) }),
+                arrayOf(false, { m: Collection<*> -> mk_Set1(*m.toTypedArray()) }),
+                arrayOf(true, { m: Collection<*> -> as_Set(m) }),
+                arrayOf(false, { m: Collection<*> -> as_Set1(m) }),
+                arrayOf(true, { m: Collection<*> -> mk_SetExtension(*m.toTypedArray()) }),
+                arrayOf(false, { m: Collection<*> -> mk_Set1Extension(*m.toTypedArray()) }),
+            )
+    }
+
+    private fun empty() = create<Any>()
```

Comment:
:question: I'm not sure what this is adding.. is it here just to avoid declaring the generic elsewhere?

you may as well have just removed the generic from create altogether e.g. 
```
private fun create(vararg m: Any) = creator.invoke(m.toList())
```

Reply:
Whilst transferring the code from "mk_Set(...)" to "create(...)" to make new sets, Alessandro and I were having trouble with types when making empty sets. This seemed to seemed to stop those issues. Your suggestion also works and is cleaner, so I will implement it

Reply:
> is it here just to avoid declaring the generic elsewhere?

Yes, pretty much. We added the generic to create to attempt to circumvent the type elision with nested sets, but then we didn't drop it when we resolved that particular issue in a different way.